### PR TITLE
Fix User roleList retrieved as empty String when there is a previous …

### LINKED ISSUE #6464: User roleList retrieved as empty String when there is a previous authentication session available.
Steps to reproduce :
- Enable the Analytics in wso2 IS [1]
- Configure secondary user store (ABC.COM)
- Configure tenant (abc.com)
- Create a user in the secondary user store and in the tenant.
- Configure the service provider and login to the application with username and password
- copy the same URL in a different tab of the browser. This time roleList returned as an empty string as it tried to retrieve the role for the user ABC.COM/ABC.COM/alexaa@abc.com

From the code base, when we tried to authenticate with the previous session it executes the logic [2]. Where we are able to see that, roleList returned with an empty string as authenticationData.getLocalUsername() return as follows:

UserstoreDomain/username@tenantDomain 

[1]https://docs.wso2.com/display/IS570/Prerequisites+to+Publish+Statistics#PrerequisitestoPublishStatistics-Step02:EnableAnalyticsinWSO2IS

[2] https://github.com/wso2-extensions/identity-data-publisher-authentication/blob/v5.1.8/components/org.wso2.carbon.identity.data.publisher.application.authentication/src/main/java/org/wso2/carbon/identity/data/publisher/application/authentication/impl/DASLoginDataPublisherImpl.java#L127

### PR DESCRIPTION
**Pull request by denuwanthi**
…authentication session available

### Proposed changes in this pull request
Resolve https://github.com/wso2/product-is/issues/6464
[List all changes you want to add here. If you fixed an issue, please
add a reference to that issue as well.]

-

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.data.publisher.application.authentication/src/main/java/org/wso2/carbon/identity/data/publisher/application/authentication/impl/DASLoginDataPublisherImpl.java [hunk 1]
```diff
@@ -130,9 +130,15 @@ protected void publishAuthenticationData(AuthenticationData authenticationData)
 
         try {
             String roleList = null;
-            if (FrameworkConstants.LOCAL_IDP_NAME.equalsIgnoreCase(authenticationData.getIdentityProviderType())) {
-                roleList = getCommaSeparatedUserRoles(authenticationData.getUserStoreDomain() + "/" + authenticationData
-                        .getUsername(), authenticationData.getTenantDomain());
+            //User roleList retrieved as empty String when there is a previous authentication session available
+            // (different tab in the same browser). When there is a previous local authentication session,
+            //authenticationData.getIdentityProviderType() will be null.
+            if (FrameworkConstants.LOCAL_IDP_NAME.equalsIgnoreCase(authenticationData.getIdentityProviderType()) || (
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.data.publisher.application.authentication/src/main/java/org/wso2/carbon/identity/data/publisher/application/authentication/impl/DASLoginDataPublisherImpl.java]: Can we refactor the complex if condition to private method "hasLocalIdPUsedForAuthentication(..)"

Comment (denuwanthi) [reply components/org.wso2.carbon.identity.data.publisher.application.authentication/src/main/java/org/wso2/carbon/identity/data/publisher/application/authentication/impl/DASLoginDataPublisherImpl.java]: fixed with https://github.com/wso2-extensions/identity-data-publisher-authentication/pull/73/commits/fbc9e667a0fdb9c0cf99dade8d797eb4c2d7d415

File: components/org.wso2.carbon.identity.data.publisher.application.authentication/src/main/java/org/wso2/carbon/identity/data/publisher/application/authentication/impl/DASLoginDataPublisherImpl.java [hunk 2]
```diff
@@ -265,4 +269,16 @@ private String getCommaSeparatedUserRoles(String userName, String tenantDomain)
         return StringUtils.EMPTY;
     }
 
+    /**
+     * Method to check if local idp was used for authentication.
+     * @param authenticationData
+     * @return boolean value to indicate if authenticated idp is local idp.
+     */
+    private boolean hasLocalIdpUsedForAuthentication(AuthenticationData authenticationData) {
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.data.publisher.application.authentication/src/main/java/org/wso2/carbon/identity/data/publisher/application/authentication/impl/DASLoginDataPublisherImpl.java:277]: This method (at least the method description) doesn't convey the message that this is an authentication which took place using an existing local session. (no actual authentication took place. Instead it's just allowing access using an existing session)

**Review by ruwanta — APPROVED** [review]

File: components/org.wso2.carbon.identity.data.publisher.application.authentication/src/main/java/org/wso2/carbon/identity/data/publisher/application/authentication/impl/DASLoginDataPublisherImpl.java (hunk 2 above)
Comment (denuwanthi) [reply components/org.wso2.carbon.identity.data.publisher.application.authentication/src/main/java/org/wso2/carbon/identity/data/publisher/application/authentication/impl/DASLoginDataPublisherImpl.java:277]: Fixed with https://github.com/wso2-extensions/identity-data-publisher-authentication/pull/73/commits/df10e1bad72a2c3392ca71df2f863d6603df1d93

Please note that this method contains check for both first time and existing sessions. Therefore I have mentioned in comment as you explained.

**Review by hasinthaindrajee — APPROVED** [review]

**Pull request merged**